### PR TITLE
added doc removal access tests

### PR DESCRIPTION
**Pull request by sridevi-15**
#### Fixes #.

- [x] Ran `flake8`
- [x] Ran `run_repo_tests.sh`

#### Changes proposed in this pull request:

- tests added for document removal behavior

### CODE REVIEW

**Comment by sridevi-15** [comment]
@hrajput89 , though I tested document remval feature with one of the combined test of replication filtering, I remove the test in this PR.  Wiil add additional verification steps on the PR of replication filtering once your submit.

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 1]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+    sg_config = params_from_base_test_setup["sg_config"]
+    cluster_config = params_from_base_test_setup["cluster_config"]
+    sg_config = params_from_base_test_setup["sg_config"]
+    db = params_from_base_test_setup["db"]
+    cbl_db = params_from_base_test_setup["source_db"]
+    sync_gateway_version = params_from_base_test_setup["sync_gateway_version"]
+
+    username = "autotest"
+    password = "password"
+    documentObj = Document(base_url)
```
Comment (hrajput89) [inline testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: variable name should follow lowercase with underscore

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 2]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+    assert cbl_ids[1] not in sg_doc_ids, "doc B exist for the user"
+
+
+@pytest.mark.listener
+@pytest.mark.replication
+def test_doc_removal_with_multipleChannels(params_from_base_test_setup, setup_customized_teardown_test):
+    """
+        @summary:
+        1. Create users in SGW with multiple channels
+            user A -> channelA,channelB, channel C;
```
Comment (hrajput89) [inline testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: little confusion is it "channelC" or "channel C". Can you make summary consistent with variable name?

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 3]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+
+    """
+    sg_db = "db"
+    sg_url = params_from_base_test_setup["sg_url"]
+    sg_admin_url = params_from_base_test_setup["sg_admin_url"]
+    sg_blip_url = params_from_base_test_setup["target_url"]
+    base_url = params_from_base_test_setup["base_url"]
+    cluster_config = params_from_base_test_setup["cluster_config"]
+    sg_config = params_from_base_test_setup["sg_config"]
+    cluster_config = params_from_base_test_setup["cluster_config"]
```
Comment (hrajput89) [inline testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: remove unused variable "cluster_config" and "sg_config"

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 4]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+    if sync_gateway_version < "2.5.0":
+        pytest.skip('This test cannnot run with sg version below 2.5.0')
+
+    c = cluster.Cluster(config=cluster_config)
+    c.reset(sg_config_path=sg_config)
+
+    channels = ["ABC", "DEF"]
+
+    sg_client = MobileRestClient()
+    authenticator = Authenticator(base_url)
```
Comment (hrajput89) [inline testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: remove unused variable "authenticator"

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 5]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+    # 1. Create 2 docs in CBL with channel ABC, DEF
+    cbl_ids = db.create_bulk_docs(2, "cbl", db=cbl_db, channels=channels)
+
+    # 2. Create users in SGW with channel ABC, DEF
+    sg_client.create_user(sg_admin_url, sg_db, username, password=password, channels=channels)
+    cookie, session_id = sg_client.create_session(sg_admin_url, sg_db, username)
+    session = cookie, session_id
+
+    # 3. push_pull replicate to SGW
+    replicator = Replication(base_url)
```
Comment (hrajput89) [inline testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: duplicate initialisation of replicator object

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 6]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+            UserB can access docB
+            UserC cannot access docC
+    """
+
+    sg_db = "db"
+    sg_url = params_from_base_test_setup["sg_url"]
+    sg_admin_url = params_from_base_test_setup["sg_admin_url"]
+    sg_blip_url = params_from_base_test_setup["target_url"]
+    base_url = params_from_base_test_setup["base_url"]
+    cluster_config = params_from_base_test_setup["cluster_config"]
```
Comment (hrajput89) [inline testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: duplicate variable declaration of "cluster_config" and "sg_config"

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 7]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+
+    channel_A = ["ABC", "DEF", "XYZ"]
+    channel_B = ["DEF"]
+    channel_C = ["XYZ"]
+
+    doc_channel_1 = ["ABC", "DEF"]
+    doc_channel_2 = ["DEF"]
+
+    sg_client = MobileRestClient()
+    authenticator = Authenticator(base_url)
```
Comment (hrajput89) [inline testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: unused variable "authenticator"

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 8]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+
+    sg_client.create_user(sg_admin_url, sg_db, username_B, password=password, channels=channel_B)
+    cookie_B, session_id_B = sg_client.create_session(sg_admin_url, sg_db, username_B)
+    session_B = cookie_B, session_id_B
+
+    sg_client.create_user(sg_admin_url, sg_db, username_C, password=password, channels=channel_C)
+    cookie_C, session_id_C = sg_client.create_session(sg_admin_url, sg_db, username_C)
+    session_C = cookie_C, session_id_C
+
+    """ 2. create docs in SGW
```
Comment (hrajput89) [inline testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: triple quote as specifically intend for doc string. Use # for each line. Also, can you make summary consistent for variable name.

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 9]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+    sg_docs = document.create_docs(doc_id_prefix='sg_docs-A', number=num_of_docs, channels=doc_channel_1)
+    sg_docs_A = sg_client.add_bulk_docs(url=sg_url, db=sg_db, docs=sg_docs, auth=session_A)
+
+    sg_docs = document.create_docs(doc_id_prefix='sg_docs-B', number=num_of_docs, channels=doc_channel_2)
+    sg_docs_B = sg_client.add_bulk_docs(url=sg_url, db=sg_db, docs=sg_docs, auth=session_B)
+
+    sg_docs = document.create_docs(doc_id_prefix='sg_docs-C', number=num_of_docs, channels=channel_C)
+    sg_docs_C = sg_client.add_bulk_docs(url=sg_url, db=sg_db, docs=sg_docs, auth=session_C)
+
+    """ 3. Verify User A(cbl_db1) can access docA and docC.
```
Comment (hrajput89) [inline testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: same as above

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py (hunk 1 above)
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py (hunk 2 above)
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py (hunk 3 above)
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py (hunk 4 above)
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py (hunk 5 above)
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py (hunk 6 above)
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py (hunk 7 above)
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py (hunk 8 above)
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py (hunk 9 above)
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 10]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+
+    for doc in sg_docs_A:
+        assert doc["id"] in doc_ids_A, "docs ids of userA does not exist in cbl db1"
+
+    for doc in sg_docs_B:
+        assert doc["id"] in doc_ids_A, "docs ids of userA does not exist in cbl db1"
+        assert doc["id"] in doc_ids_B, "docs ids of userB does not exist in cbl db2"
+
+    for doc in sg_docs_C:
+        assert doc["id"] not in doc_ids_A, "docs ids of userA does not exist in cbl db1"
```
[the remark this answers is not in the document]
Comment (euniceatcouchbase) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: to my understanding, seems cbl_db1 is matching to userA, so as doc_ids_A. The assertion only check which user expects to access what data, however, there is lack of opposite assertions, how about:  doc in sg_docs_C => assert doc[id] not in doc_ids_C? does it translate to userC cannot access docC? or I can be totally wrong

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 11]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+    base_url = params_from_base_test_setup["base_url"]
+    cluster_config = params_from_base_test_setup["cluster_config"]
+    sg_config = params_from_base_test_setup["sg_config"]
+    cluster_config = params_from_base_test_setup["cluster_config"]
+    sg_config = params_from_base_test_setup["sg_config"]
+    db = params_from_base_test_setup["db"]
+    cbl_db = params_from_base_test_setup["source_db"]
+    sync_gateway_version = params_from_base_test_setup["sync_gateway_version"]
+
+    username = "autotest"
```
[the remark this answers is not in the document]
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: Sure , we can use that. Let's have another PR for that. This PR is specific to one of the feature, let's don't combine with feature PRs.

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py [hunk 12]
```diff
@@ -3215,6 +3215,260 @@ def test_replication_push_replication_guest_enabled(params_from_base_test_setup,
...
+    replicator = Replication(base_url)
+    authenticator = Authenticator(base_url)
+    replicator_authenticator = authenticator.authentication(session_id, cookie, authentication_type="session")
+    repl = replicator.configure_and_replicate(source_db=cbl_db,
+                                              target_url=sg_blip_url,
+                                              continuous=True,
+                                              replicator_authenticator=replicator_authenticator)
+    # replicator.stop(repl)
+
+    # 4. remove doc A from channel A
```
[the remark this answers is not in the document]
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: yes

File: testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sridevi-15) [reply testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py]: I already have that assertion for doc[id] not in doc_ids_C inside the last 'for' loop. However, I noticed  that the assertion message is wrong , I have edited the message and committed.

**Pull request merged**